### PR TITLE
fix the out of memory erro when gtbbox is too much

### PR DESCRIPTION
**Pull request by ziming-liu**
if we custom a new dataset, if the gt boxes of one image up to 300 or  more, we often get  GPU out of memory erro .  i find that the reason coming from the computation of  IOU part,  so  we can turn the operation to cpu is  gtbbox is too much and  come back to GPU later.

### CODE REVIEW

**Comment by hellock** [comment]
Thanks for the PR! Could you fix the linting errors? We use flake8 as the linter.

**Comment by hellock** [comment]
#188

**Comment by wei-yuma** [comment]
> 
> 
> if we custom a new dataset, if the gt boxes of one image up to 300 or more, we often get GPU out of memory erro . i find that the reason coming from the computation of IOU part, so we can turn the operation to cpu is gtbbox is too much and come back to GPU later.

sorry your code  seems wrong , to let a tensor be moved to gpu should do `tensor.cuda()`, not `tensor.gpu()` and, your code doesn`t fix the out of memory erro when gtbbox is too much

**Comment by ziming-liu** [comment]
> > if we custom a new dataset, if the gt boxes of one image up to 300 or more, we often get GPU out of memory erro . i find that the reason coming from the computation of IOU part, so we can turn the operation to cpu is gtbbox is too much and come back to GPU later.
> 
> sorry your code seems wrong , to let a tensor be moved to gpu should do `tensor.cuda()`, not `tensor.gpu()` and, your code doesn`t fix the out of memory erro when gtbbox is too much

sorry, i run the modified code on my another copy,  and im sure it's OK to reduce OOM by moving the operation of  computing  IOU  to cup().  but I  write some code wrongly  in this pull request.

**Comment by ziming-liu** [comment]
> > if we custom a new dataset, if the gt boxes of one image up to 300 or more, we often get GPU out of memory erro . i find that the reason coming from the computation of IOU part, so we can turn the operation to cpu is gtbbox is too much and come back to GPU later.
> 
> sorry your code seems wrong , to let a tensor be moved to gpu should do `tensor.cuda()`, not `tensor.gpu()` and, your code doesn`t fix the out of memory erro when gtbbox is too much

I run the code again, and fix some problem. And  I think the  OOM problem mainly due to the  operation of assign. so  I try to turn the assign operation to cpu.  In my dataset, which have 20--300 gtbbox per image but  some images up to 900 gtbboxes per image,  the method  works in solving OOM.  If you think this is not true, please let me know , thank you !

**Comment by wei-yuma** [comment]
@ziming-liu yeah, It seems right

File: mmdet/core/bbox/assigners/max_iou_assigner.py [hunk 1]
```diff
@@ -88,6 +97,12 @@ def assign(self, bboxes, gt_bboxes, gt_bboxes_ignore=None, gt_labels=None):
             overlaps[:, ignore_max_overlaps > self.ignore_iof_thr] = -1
 
         assign_result = self.assign_wrt_overlaps(overlaps, gt_labels)
+        # trun the assign results back to cuda
```
Comment (hellock) [inline mmdet/core/bbox/assigners/max_iou_assigner.py]: A typo here: trun -> turn.
"move" or "convert" may be better.

File: mmdet/core/bbox/assigners/max_iou_assigner.py [hunk 2]
```diff
@@ -72,6 +72,15 @@ def assign(self, bboxes, gt_bboxes, gt_bboxes_ignore=None, gt_labels=None):
         """
         if bboxes.shape[0] == 0 or gt_bboxes.shape[0] == 0:
             raise ValueError('No gt or bboxes')
+        # trun the assign op to cup
+        if gt_bboxes.shape[0] > 0:
```
[the remark this answers is not in the document]
Comment (ziming-liu) [reply mmdet/core/bbox/assigners/max_iou_assigner.py]: yes, I agree, it's much better to set a flag in configfile.

File: mmdet/core/bbox/transforms.py [hunk 3]
```diff
@@ -28,7 +28,7 @@ def bbox2delta(proposals, gt, means=[0, 0, 0, 0], stds=[1, 1, 1, 1]):
     stds = deltas.new_tensor(stds).unsqueeze(0)
     deltas = deltas.sub_(means).div_(stds)
 
-    return deltas
+    return deltas.cuda()
```
[the remark this answers is not in the document]
Comment (ziming-liu) [reply mmdet/core/bbox/transforms.py]: oh no it's not necessary anymore

File: mmdet/core/bbox/assigners/max_iou_assigner.py (hunk 1 above)
Comment (ziming-liu) [reply mmdet/core/bbox/assigners/max_iou_assigner.py]: ok I have made the change

**Comment by hellock** [comment]
It seems that you pushed too much commits, including sensitive data.

**Comment by hellock** [comment]
Fixed in #1575.

**Pull request closed without merging**